### ISSUE TITLE
Investigation: try React-Redux v7 / v8 with React 18 alpha

### ISSUE THREAD

**Issue by markerikson** [reduxjs/react-redux#1732]
React 18 is now available as preview alpha releases (per https://reactjs.org/blog/2021/06/08/the-plan-for-react-18.html ).

The React team has put together a Working Group discussion forum to describe the changes in React 18 in more detail: https://github.com/reactwg/react-18/discussions

Looking at that forum, there's several issues that may be relevant to React-Redux in some way:

- Overviews:
  - [Introducing React 18](https://github.com/reactwg/react-18/discussions/4)
  - [Upgrading to React 18 on the Client](https://github.com/reactwg/react-18/discussions/6)
  - [Installing React 18 Alpha](https://github.com/reactwg/react-18/discussions/9)
- Behavior changes:
  - [Automatic batching](https://github.com/reactwg/react-18/discussions/21)
  - [Adding Strict Effects to StrictMode](https://github.com/reactwg/react-18/discussions/19)
  - [How to support Strict Effects](https://github.com/reactwg/react-18/discussions/18)
  - [Testing with Strict Effects](https://github.com/reactwg/react-18/discussions/17)
  - [How to support Strict Effects](https://github.com/reactwg/react-18/discussions/18)

I'd like to have some folks start trying out React-Redux v7 with React 18 alpha, and report any interesting findings.  We could potentially use some of the example apps from the Redux tutorials, and any of the example CodeSandboxes linked in our docs.

You would need to:

- Switch the React version to one of the React 18 alphas, and update the app to use `React.createRoot()` per instructions
- Try out the app
- Hand-inspect the behavior to see if anything seems unexpected

We're interested in seeing examples and comparisons with both `connect` and `useSelector`.

It would also be worth cloning the React-Redux repo, and updating our React devDep to run our tests against the alphas.

The React team has reported that a brief test of a couple Redux-based apps combined with `createRoot + StrictMode` did not appear to be working correctly - in particular, no re-rendering happening despite actions being dispatched.

It's very possible that Strict Effects running effects callbacks more than once is causing problems with subscriptions.

We'd appreciate any feedback reports!  

**Please provide this info**:

- Details on the example app used
- Link to updated source code / CodeSandbox if possible
- Description of observed behavior, good or bad ("works okay", "clicking button X doesn't update the UI", etc)

**Comment by gaearon** [comment]
Note to folks trying — if you have StrictMode on, please test *both* with StrictMode and without it. This is to see if some issues only occur in StrictMode.

**Comment by timdorr** [comment]
We actually have tests for StrictMode in our suite already, so it should be good to test at least the basics here.

**Comment by markerikson** [comment]
Okay.  First issue here is that there's this new "Strict Effects" thing which is enabled if you're using `createRoot()` + `<StrictMode>`, which double-runs the effects in the same way that `<StrictMode>` has traditionally double-run renders.  I could easily see that messing up subscriptions.

So, we'd need to figure out how to do a React 18 upgrade for our tests, which I assume would mean something with React Testing Library since that's doing the rendering?

**Comment by nickmccurdy** [comment]
If you're asking about Testing Library compatibility, you may want to subscribe to https://github.com/testing-library/react-testing-library/pull/925.

**Comment by ryota-murakami** [comment]
I have a side project that used to React-Redux v7.2.4, I'd like to try it if React 18 alpha working on CRA v4.0.3.

**Comment by markerikson** [comment]
Comment from the React team on plans for helping the community do updates:

https://github.com/reactwg/react-18/discussions/56#discussioncomment-865019

> The first step is to explain what work we see needs to be done at a high level, so we're working on a workgroup post that will be a "Concurrent rendering for library maintainers" guide, and documentation for new APIs libraries will use to support concurrent features. Those docs will be a "kick off" for starting library support. After they're published, we'll do a Q&A / support session, and use the working group to document questions, use cases, patterns, design, etc.
> 
> I know we're all really excited to start with the work of adding support to libraries, so I'm sorry it's taking so long. We've spent the last couple weeks kicking off the release (which focused on the gradual adoption strategy), and now we're transitioning to working with libraries to support the new concurrent features.
> 
> We expect this work to take months, not weeks, similar to how the hooks release went.

**Comment by markerikson** [comment]
Mobx has a PR going that's tracking some similar status checks for React 18. Worth keeping an eye on that:

https://github.com/mobxjs/mobx/pull/3005

**Comment by artem-malko** [comment]
Hi @markerikson! I have something interesting for you) 

I have a [repository](https://github.com/artem-malko/react-ssr-template) with react 18 + SSR + redux for UI state. There is an interactive demo — http://158.101.223.0:5000/

As you can see, there is a button "Patch query string". If you'll click this button, the UI state in a redux-store will be updated with new Date().toString() There is [a source code](https://github.com/artem-malko/react-ssr-template/blob/main/src/ui/main/app.tsx#L22). Top part of the page has a component, that has [a subscription to the redux-store vis useSelector](https://github.com/artem-malko/react-ssr-template/blob/main/src/ui/main/app.tsx#L19).

So, if you will open the demo with a query param strict=true, the [whole app will be wrapped](https://github.com/artem-malko/react-ssr-template/blob/main/src/applications/client/index.tsx#L34) with <StrictMode> component. If there is no strict query param — without it. Click to the button, that changes the query string works perfectly in strict mode and without it. But, **it works in production build only**. 

If you will clone the repository, exec `npm i && make dev` and open http://localhost:4000?strict=true, you will see, that the. UI won't be changed after the button click. 

Actually, I can not understand, what is the difference between develop and production build in case of redux/react-redux. May be it is because of react?

**Comment by gaearon** [comment]
Example of StrictMode + createRoot breaking it, per request from https://github.com/reactwg/react-18/discussions/19#discussioncomment-1320450: https://codesandbox.io/s/serverless-wood-uouyy?file=/src/index.js

**Comment by markerikson** [comment]
And copying over some excellent detective work from @andarist :

> The problem is not that dispatching doesn't work (although you probably didn't mean exactly that) and it's not even because components are not subscribed - because they are. The problem is that the hooked up listener is a "proxy" listener:
> https://github.com/reduxjs/react-redux/blob/d859c236168ddf53a97a032ac3d098ee670b8a40/src/utils/Subscription.ts#L111-L115
> which actually forwards notifications to another one that is put as a property on the subscription object. And that property gets "nullified" in the cleanup of the `Provider`:
> https://github.com/reduxjs/react-redux/blob/d859c236168ddf53a97a032ac3d098ee670b8a40/src/components/Provider.tsx#L43
> while not being re-instantiated in the setup because it's only assigned within `useMemo` here:
> https://github.com/reduxjs/react-redux/blob/d859c236168ddf53a97a032ac3d098ee670b8a40/src/components/Provider.tsx#L25
> 
> So what happens is that Provider gets notified about a store change - it should notify nested subscriptions but it doesn't because of the mentioned issue. And `useSelector` (and probably any other inner subscribers like `connect` etc) are never notified about this because they are subscribed to that parent subscription that lives in the `Provider` here.
> 
> TLDR; lack of symmetry between setup/cleanup of an effect in the `Provider` led to the notification chain being broken.

**Comment by serprex** [comment]
https://etg.dek.im works okay after updating to react-redux 8.0.0-beta.3 & react 18.0

**Comment by markerikson** [comment]
React-Redux v8 is out, so I think we can call this done :)
